cmd/kubeadm/app/util/config/cluster.go

```diff
@@ -180,23 +182,61 @@ func getNodeNameFromKubeletConfig(kubeconfigDir string) (string, error) {
 	return strings.TrimPrefix(cert.Subject.CommonName, constants.NodesUserPrefix), nil
 }
 
-// getAPIEndpoint returns the APIEndpoint for the current node
-func getAPIEndpoint(data map[string]string, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint) error {
-	// gets the ClusterStatus from kubeadm-config
-	clusterStatus, err := UnmarshalClusterStatus(data)
+func getAPIEndpoint(client clientset.Interface, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint) error {
+	if err := getAPIEndpointFromPodAnnotationWithRetry(client, nodeName, apiEndpoint); err == nil {
+		return nil
+	}
+	if err := getAPIEndpointFromClusterStatus(client, nodeName, apiEndpoint); err == nil {
+		return nil
+	}
+	return errors.Errorf("could not retrieve API endpoint for node name %q", nodeName)
+}
+
+func getAPIEndpointFromPodAnnotationWithRetry(client clientset.Interface, nodeName string, apiEndpoint *kubeadmapi.APIEndpoint) error {
+	var rawAPIEndpoint string
+	var lastErr error
+	err := wait.ExponentialBackoff(clientsetretry.DefaultRetry, func() (bool, error) {
+		rawAPIEndpoint, lastErr = getRawAPIEndpointFromPodAnnotation(client, nodeName)
+		return lastErr == nil, nil
+	})
 	if err != nil {
 		return err
 	}
+	parsedAPIEndpoint, err := kubeadmapi.APIEndpointFromString(rawAPIEndpoint)
+	if err != nil {
+		return errors.Wrapf(err, "could not parse API endpoint for node %q", nodeName)
+	}
+	*apiEndpoint = parsedAPIEndpoint
+	return nil
+}
 
-	// gets the APIEndpoint for the current machine from the ClusterStatus
-	e, ok := clusterStatus.APIEndpoints[nodeName]
-	if !ok {
-		return errors.New("failed to get APIEndpoint information for this node")
+func getRawAPIEndpointFromPodAnnotation(client clientset.Interface, nodeName string) (string, error) {
+	podList, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(
+		metav1.ListOptions{
+			FieldSelector: fmt.Sprintf("spec.nodeName=%s", nodeName),
+			LabelSelector: fmt.Sprintf("component=%s,tier=%s", constants.KubeAPIServer, constants.ControlPlaneTier),
+		},
+	)
+	if err != nil {
+		return "", errors.Wrap(err, "could not retrieve list of pods to determine api server endpoints")
 	}
+	if len(podList.Items) == 0 {
+		return "", errors.Errorf("could not find API server pod for node name %q", nodeName)
+	}
+	return podList.Items[0].Annotations[constants.KubeAPIServerAdvertiseAddressEndpointAnnotationKey], nil
```

Comment:
This is a certain crash here if we find an API server pod, that was created by an old version of kubeadm and does not have the annotation. We must check for the annotation key presence. This would have been caught with the master upgrade job after merge. The reason it's not in the PR, is that there are no unit tests to these funcs.

Another thing here. Everybody with an admin account here can intentionally or accidentally create a pod that has the required labels and is on the same node. Therefore, we should probably error out if more than one pod is found on the returned list.

Reply:
> This is a certain crash here if we find an API server pod, that was created by an old version of kubeadm and does not have the annotation. We must check for the annotation key presence. This would have been caught with the master upgrade job after merge. The reason it's not in the PR, is that there are no unit tests to these funcs.

Yes, this is right, will update.

> Another thing here. Everybody with an admin account here can intentionally or accidentally create a pod that has the required labels and is on the same node. Therefore, we should probably error out if more than one pod is found on the returned list.

On the case of more than one pod I was thinking in iterating over the N pods and only error out if we get different annotations (if we have N pods, but all have the same annotation I would not error out). But thought that as a first approach the simplest solution would be better. I can update the PR.